smqtk_detection/interfaces/object_detector.py

```diff
@@ -120,7 +123,10 @@ def detect_objects(
             yield de
 
     @abc.abstractmethod
-    def _detect_objects(self, data: DataElement) -> Iterator[Tuple[AxisAlignedBoundingBox, Dict[Hashable, float]]]:
+    def _detect_objects(
+            self,
+            data: DataElement
+    ) -> Optional[Iterator[Tuple[AxisAlignedBoundingBox, Dict[Hashable, float]]]]:
```

Comment:
(oops, I meant to put this on the `detect_objects` method, but you know what I mean)

Reply:
This issues comes back to [object_detector.py#L314](https://github.com/Kitware/SMQTK-Detection/blob/e0529136a0f0de7f6a2649a6fc34e9aa98039c1d/smqtk_detection/interfaces/object_detector.py#L314).

`smqtk_image_io.interfaces.image_reader.ImageReader.load_as_matrix` can return a `None` which means the return is `Optional`. It requires `Optional` because of the `smqtk_image_io.interfaces.image_reader.ImageReader._get_matrix_property` can return `None` and is `Optional`.

Since `load_as_matrix` can return `None`, either `_detect_objects_matrix` must accept `None` as a parameter or the function `_detect_objects` must be able to return `None`. It doesn't make sense to `detect_objects` of a `None`, but it does make sense that `detect_objects` could return `None` for no objects detected. So the addition of `Optional` to these returned values is to reflect `None` (no detections) up the call stack.

Subsequently, there was a change made to iterating over the return of `_detect_objects` because `None` is not an iterator. Since we have no detections to iterator over, that must also return `None` for no detections.